backend/support/cog_over_w_audit.py

```diff
@@ -0,0 +1,289 @@
+from collections import defaultdict
+import logging
+import math
+
+from django.db.models.functions import Cast
+from django.db.models import BigIntegerField, Q
+
+from support.models import CognizantAssignment
+from dissemination.models import MigrationInspectionRecord
+from django.apps import apps
+
+logger = logging.getLogger(__name__)
+
+
+COG_LIMIT = 50_000_000
+DA_THRESHOLD_FACTOR = 0.25
+
+FIRST_BASELINE_YEAR = 2019
+BASELINE_VALID_FOR_YEARS = 5
+
+DBKEY_TO_UEI_TRANSITION_YEAR = "2022"
+
+
+def compute_cog_over(
+    federal_awards, submission_status, auditee_ein, auditee_uei, audit_year
+):
+    """
+    Compute cog or oversight agency for the sac.
+    Return tuple (cog_agency, oversight_agency)
+
+    WIP:
+    - sac.federal_awards
+    - sac.submission_status
+    - sac.ein
+    - sac.auditee_uei
+    """
+    if not federal_awards:
+        logger.warning(
+            f"Trying to determine cog_over for a sac with zero awards with status = {submission_status}."
+        )
+        return (None, None)
+    awards = federal_awards["awards"]
+    total_amount_expended = federal_awards["total_amount_expended"]
+    cognizant_agency = oversight_agency = None
+    (max_total_agency, max_da_agency) = calc_award_amounts(awards)
+
+    agency = determine_agency(
+        total_amount_expended,
+        max_total_agency,
+        max_da_agency,
+    )
+
+    if total_amount_expended <= COG_LIMIT:
+        oversight_agency = agency
+        # logger.warning("Assigning an oversight agenct", oversight_agency)
+        return (cognizant_agency, oversight_agency)
+
+    cognizant_agency = determine_hist_agency(auditee_ein, auditee_uei, audit_year)
+    if cognizant_agency:
+        return (cognizant_agency, oversight_agency)
+    cognizant_agency = agency
+    return (cognizant_agency, oversight_agency)
```

Comment:
```suggestion
    return (cognizant_agency if cognizant_agency else agency, oversight_agency)
```

This _may_ be a bit simpler, there are a whole lot of return statements in this method, which adds a lot of complexity/code paths. If we can clean it up some that would be great.

Reply:
This is not in the scope of SOT.

Reply:
To be addressed in https://github.com/GSA-TTS/FAC/issues/4799